### PR TITLE
datavalues2: Refactor all string2number functions

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Part of https://github.com/datafuselabs/databend/issues/3975

In this PR, I refactor the whole `String2NumberFunction` to make it works well with datavalue2.

## Changelog

- New Feature

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Review by sundy-li — APPROVED** [review]

**Review by junli1026 — APPROVED** [review]

**Comment by Xuanwo** [comment]
`inet_ntoa` seems wrong?

**Comment by Xuanwo** [comment]
> `inet_ntoa` seems wrong?

Oh, it's a silly typo, fixed.

**Comment by BohuTANG** [comment]
Along the work of #4074 , is this PR still available? @sundy-li

**Comment by sundy-li** [comment]
Let's merge this after #4074

**Comment by sundy-li** [comment]
@Xuanwo  Pls merge this pr from main and send another pr into main.

**Comment by Xuanwo** [comment]
With the help from Github, we can switch `base` directly without sending new PRs. Let's keep working on this one.

**Comment by Xuanwo** [comment]
I force pushed all commits, so that we have a cleaner history.

**Comment by PsiACE** [comment]
```
[ FAIL ] - result differs with:
--- /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0023_function_strings_ascii.result	2022-02-09 06:43:13.599899514 +0000
+++ /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0023_function_strings_ascii.stdout	2022-02-09 06:46:10.992957217 +0000
@@ -1,6 +1,6 @@
+ERROR 1105 (HY000) at line 1: Code: 1007, displayText = Expected string or null, but got UInt8 (while in select before projection).
 49
+ERROR 1105 (HY000) at line 3: Code: 1007, displayText = Expected string or null, but got UInt8 (while in select before projection).
 49
-49
-49
-45
+ERROR 1105 (HY000) at line 5: Code: 1007, displayText = Expected string or null, but got Int8 (while in select before projection).
 45
```

```
[ FAIL ] - result differs with:
--- /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0030_function_strings_octet_length.result	2022-02-09 06:43:13.603899534 +0000
+++ /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0030_function_strings_octet_length.stdout	2022-02-09 06:46:12.320965236 +0000
@@ -1,6 +1,6 @@
+ERROR 1105 (HY000) at line 1: Code: 1007, displayText = Expected string or null, but got UInt8 (while in select before projection).
 2
-2
-3
+ERROR 1105 (HY000) at line 3: Code: 1007, displayText = Expected string or null, but got UInt16 (while in select before projection).
 3
 5
 7
```

```
[ FAIL ] - result differs with:
--- /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0036_function_strings_ord.result	2022-02-09 06:43:13.603899534 +0000
+++ /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0036_function_strings_ord.stdout	2022-02-09 06:46:14.228976757 +0000
@@ -4,8 +4,4 @@
 4036989590
 97
 4036989590
-48
-49
-50
-51
-52
+ERROR 1105 (HY000) at line 7: Code: 1007, displayText = Expected string or null, but got UInt64 (while in select before order by).
```

File: common/functions/src/scalars/strings/string2number.rs [hunk 1]
```diff
@@ -64,44 +61,53 @@ where
...
         }
-        let nullable = args.iter().any(|arg| arg.is_nullable());
-        let dt = T::return_type()?;
-        Ok(DataTypeAndNullable::create(&dt, nullable))
     }
 
-    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
-        let column = columns[0]
-            .column()
-            .cast_with_type(&DataType::String)?
```
Comment (Xuanwo) [inline common/functions/src/scalars/strings/string2number.rs:98]: @sundy-li It seems that I need to input into `String`? I'm meeting errors like

```
2022-02-09T09:34:07.340102Z ERROR databend_query::servers::mysql::writers::query_result_writer: OnQuery Error: Code: 1058, displayText = downcast column error, column type: UInt64, expected column: "common_datavalues2::columns::string::StringColumn".
```

logs: https://github.com/datafuselabs/databend/runs/5122838220?check_suite_focus=true#step:3:1761

Comment (sundy-li) [reply common/functions/src/scalars/strings/string2number.rs:98]: Yes, I think we should forbid the implicit cast in `return_type`, so some tests should be removed.

1. return_type just check the args[0] is `string` or not.

2. `passthrough_null` is default to true, so we do not need to check null types.


In `function2_adaptor`:
```
  if self.passthrough_null() {
            // one is null, result is null
            if args.iter().any(|v| v.data_type_id() == TypeID::Null) {
                return Ok(NullType::arc());
            }
```

**Comment by BohuTANG** [comment]
@mergify update

**Pull request merged**